server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java

```diff
@@ -723,8 +727,287 @@ public void testDesiredBalanceShouldConvergeInABigCluster() {
+                        )
+                        .build();
+                    indices++;
+                    change = "create index ["
+                        + indexMetadata.getIndex().getName()
+                        + "] with total ["
+                        + (shards * (replicas + 1))
+                        + "] shards";
+                }
+                case 1 -> {
+                    var allIndices = clusterState.metadata().indices().keySet().stream().filter(name -> name.startsWith("index-")).toList();
+                    if (allIndices.isEmpty() == false) {
+                        var indexToGrow = randomFrom(allIndices);
+                        var originalSize = clusterDataSizes.getOriginalIndexShardSize(indexToGrow);
+                        clusterDataSizes.set(clusterState.metadata().index(indexToGrow), smallShardSizeDeviation(originalSize, 10));
+                        change = "change size of the index " + indexToGrow;
+                    }
+                }
+                case 2 -> {
+                    var indexToDelete = clusterState.metadata()
+                        .indices()
+                        .keySet()
+                        .stream()
+                        .filter(name -> name.startsWith("index-"))
+                        .sorted()
+                        .min(CharSequence::compare);
+
+                    if (indexToDelete.isPresent()) {
+                        clusterDataSizes.remove(clusterState.metadata().index(indexToDelete.get()));
+                        clusterState = ClusterState.builder(clusterState)
+                            .metadata(Metadata.builder(clusterState.metadata()).remove(indexToDelete.get()).build())
+                            .routingTable(RoutingTable.builder(clusterState.routingTable()).remove(indexToDelete.get()).build())
+                            .build();
+                        change = "delete index " + indexToDelete.get();
+                    }
+                }
+            }
+
+            var previousDesiredBalance = desiredBalance;
+            desiredBalance = desiredBalanceComputer.compute(
+                previousDesiredBalance,
+                new DesiredBalanceInput(
+                    computation.getAndIncrement(),
+                    routingAllocationWithDecidersOf(clusterState, clusterDataSizes.getClusterInfo(clusterState), Settings.EMPTY),
+                    List.of()
+                ),
+                queue(),
+                ignored -> true
+            );
+
+//            var unassigned = desiredBalance.assignments().entrySet().stream().filter(entry -> entry.getValue().unassigned() > 0).toList();
+//            assertThat(unassigned, hasSize(0));
+
+            var moves = DesiredBalance.shardMovements(previousDesiredBalance, desiredBalance);
+            // TODO compute moved data size
+            logger.info("Computed {} moves after mutation: {}", moves, change);
+            // TODO set up quality barrier
```

Comment:
I am still thinking what could be a quality gate here.
May be we should not rebalance more shards then we just added or removed by creating or deleting an index? Or may be limit rebalanced shard sizes in proportion to overall cluster size?